tests/support/client.rs

```diff
@@ -249,36 +294,129 @@ impl hyper::client::connect::Connect for Conn {
     }
 }
 
-/// A wrapper around a TcpStream, allowing us to signal when the connection
-/// is dropped.
-struct RunningIo {
-    inner: TcpStream,
-    /// When this drops, the related Receiver is notified that the connection
-    /// is closed.
-    running: oneshot::Sender<()>,
+enum ConnectorFuture {
+    Init {
+        future: tokio::net::tcp::ConnectFuture,
+        tls: Option<TlsConfig>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
+    Handshake {
+        future: tokio_rustls::Connect<TcpStream>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
 }
 
-impl io::Read for RunningIo {
+impl Future for ConnectorFuture {
+    type Item = RunningIo;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let take_running = move |r: &Mutex<Option<oneshot::Sender<()>>>| {
+            r.lock()
+                .expect("running lock")
+                .take()
+                .expect("support client cannot connect more than once")
+        };
+
+        loop {
+            *self = match self {
+                ConnectorFuture::Init {
+                    future,
+                    tls,
+                    running,
+                } => {
+                    let io = try_ready!(future.poll());
+
+                    match tls {
+                        None => {
+                            return Ok(Async::Ready(RunningIo::Http(io, take_running(running))));
+                        }
+
+                        Some(TlsConfig {
+                            client_config,
+                            name,
+                        }) => {
+                            let future = tokio_rustls::TlsConnector::from(client_config.clone())
+                                .connect(DNSName::as_ref(name), io);
+                            ConnectorFuture::Handshake {
+                                future,
+                                running: running.clone(),
+                            }
+                        }
+                    }
+                }
+
+                ConnectorFuture::Handshake { future, running } => {
+                    let io = try_ready!(future.poll());
+                    return Ok(Async::Ready(RunningIo::Https(io, take_running(running))));
+                }
+            }
+        }
+    }
+}
+
+enum RunningIo {
+    Http(TcpStream, oneshot::Sender<()>),
+    Https(TlsStream<TcpStream, ClientSession>, oneshot::Sender<()>),
+}
+
+impl Read for RunningIo {
+    #[inline]
```

Comment:
Is there a specific reason you added the `#[inline]` attributes? If not, I'm inclined to suggest we leave them out. If we start addressing test performance, there are probably some larger module changes we could make before worrying about hinting inlines to the compiler!